File: benchmark/src/main/java/de/hhu/bsinfo/observatory/benchmark/RdmaLatencyOperation.java
package de.hhu.bsinfo.observatory.benchmark;

import de.hhu.bsinfo.observatory.benchmark.Benchmark.Mode;
import de.hhu.bsinfo.observatory.benchmark.Benchmark.RdmaMode;
import de.hhu.bsinfo.observatory.benchmark.result.Status;

public class RdmaLatencyOperation extends LatencyOperation {

    private final RdmaMode rdmaMode;

    RdmaLatencyOperation(Benchmark benchmark, Mode mode, int operationCount, int operationSize, RdmaMode rdmaMode) {
        super(benchmark, mode, operationCount, operationSize);

        this.rdmaMode = rdmaMode;
    }

    @Override
    boolean needsFilledReceiveQueue() {
        return false;
    }

    @Override
    Status warmUp(int operationCount) {
        if(getMode() == Mode.SEND) {
            for(int i = 0; i < operationCount; i++) {
                Status status = getBenchmark().performSingleRdmaOperation(rdmaMode);

                if(status != Status.OK) {
                    return status;
                }
            }
        }

        getBenchmark().synchronize();

        return Status.OK;
    }

    @Override
    public Status execute() {
        Status status = Status.OK;

        if(getMode() == Mode.SEND) {
            for(int i = 0; i < getMeasurement().getOperationCount(); i++) {
                getMeasurement().startSingleMeasurement();
                status = getBenchmark().performSingleRdmaOperation(rdmaMode);
                getMeasurement().stopSingleMeasurement();

                if(status != Status.OK) {
                    return status;
                }
            }

            getMeasurement().finishMeasuring();
        }

        getBenchmark().synchronize();

        return status;
    }
}

File: cpp/src/benchmark/operation/RdmaLatencyOperation.cpp
#include "RdmaLatencyOperation.h"

namespace Observatory {

RdmaLatencyOperation::RdmaLatencyOperation(Benchmark *benchmark, Benchmark::Mode mode, uint32_t operationCount, uint32_t operationSize, Benchmark::RdmaMode rdmaMode) :
        LatencyOperation(benchmark, mode, operationCount, operationSize),
        rdmaMode(rdmaMode) {}

const char *RdmaLatencyOperation::getClassName() const {
    return "RdmaLatencyOperation";
}

bool RdmaLatencyOperation::needsFilledReceiveQueue() const {
    return getMode() != Benchmark::Mode::SEND;
}

Status RdmaLatencyOperation::warmUp(uint32_t operationCount) {
    Status status = Status::OK;

    if(getMode() == Benchmark::Mode::SEND) {
        for(uint32_t i = 0; i < operationCount; i++) {
            status = getBenchmark().performSingleRdmaOperation(rdmaMode);

            if(status != Status::OK) {
                return status;
            }
        }
    }

    if(!getBenchmark().synchronize()) {
        return Status::SYNC_ERROR;
    }

    return status;
}

Status RdmaLatencyOperation::execute() {
    Status status = Status::OK;

    if(getMode() == Benchmark::Mode::SEND) {
        auto startTime = std::chrono::steady_clock::now();

        for(uint32_t i = 0; i < getMeasurement().getOperationCount(); i++) {
            getMeasurement().startSingleMeasurement();
            status = getBenchmark().performSingleRdmaOperation(rdmaMode);
            getMeasurement().stopSingleMeasurement();

            if(status != Status::OK) {
                return status;
            }
        }

        uint64_t time = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now() - startTime).count();
        getMeasurement().finishMeasuring(time);
    }

    if(!getBenchmark().synchronize()) {
        return Status::SYNC_ERROR;
    }

    return status;
}

}